File: 2022/day02.py
import fileinput
from collections import defaultdict, Counter

lines = [line.strip() for line in fileinput.input()]
rock = {'C': 'S', 'Z': "S", 'A': 'R', 'B': 'P',  'X': 'R', 'Y': 'P'}
beats = {'R': 'S', 'P': 'R', 'S': 'P'}
loses = {'R': 'P', 'P': 'S', 'S': 'R'}
pts = {'R': 1, 'P': 2, 'S': 3}
silver = 0
gold = 0

for line in lines:
    l, r = line.split()
    l = rock[l]
    r = rock[r]
    if r == 'R':
        rg = beats[l]
    elif r == 'P':
        rg = l
    else:
        rg = loses[l]
    silver += (6 if beats[r] == l else (l == r) * 3) + pts[r]
    gold += (6 if beats[rg] == l else (l == rg) * 3) + pts[rg]
print('silver', silver)
print('gold', gold)

File: 2022/src/day02.rs
use std::collections::HashMap;

fn points(l: u8, r: u8) -> u8 {
    if l == r {
        3 + r
    } else if l == 1 + (r + 1) % 3 {
        6 + r
    } else {
        r
    }
}

pub fn main() {
    let contents = include_str!("../2");
    let lines = contents
        .lines()
        .map(|set| set.split(' ').collect::<Vec<_>>())
        .collect::<Vec<_>>();
    let nxt = HashMap::from([("A", " CAB"), ("B", " ABC"), ("C", " BCA")]);
    let score = HashMap::from([("A", 1), ("B", 2), ("C", 3), ("X", 1), ("Y", 2), ("Z", 3)]);

    let mut silver: i32 = 0;
    let mut gold: i32 = 0;
    for v in lines {
        let (l, r) = (v[0], v[1]);
        silver += points(score[l], score[r]) as i32;
        gold += points(
            score[l],
            score[nxt[l]
                .chars()
                .nth(score[r] as usize)
                .unwrap()
                .to_string()
                .as_str()],
        ) as i32;
    }

    println!("silver: {}", silver);
    println!("gold: {}", gold);
}
